Classfile API stack map processing

asotona opened the pull request.

briangoetz commented:
In reviewing this PR, I noticed a vestige of an earlier abandoned 
direction, which is CodeElement.Kind.  There are Kind values for 
PARAMETER_ANNOTATION (basically unused), TYPE_ANNOTATION (misused), and 
you added one for STACK_MAP.  I think all three of these should be 
replaced with a single ATTRIBUTE kind, and the codeKind() of any 
attribute that extends CodeElement should be ATTRIBUTE.  The END kind is 
also effectively dead and should be removed (it is used only by a dead 
opcode.)

This vestige is also present in your stack map attribute:

         @Override
         public Opcode opcode() {
             return Opcode.TYPE_ANNO;
         }

which is clearly the wrong opcode to return for stack maps, but its also 
the wrong thing to return for type annotation attributes as well.  
Probably best to create a pseudo opcode for ATTRIBUTE to return in these 
places too.

My reading of this PR is that:

  - You add a new option, PROCESS_STACK_MAP
  - The behavior of PSM is to deliver the StackMap attribute whole when 
iterating the code elements (as we do with e.g. RVTA).
  - If SM generation has been disabled, and the user has send a stack 
map attribute to the builder, we can write that (?)

I like that you have to disengage the automatic stack map generation in 
order to be able to generate your own, but this last bullet still makes 
me a little uncomfortable.  I'll think about how to make this more clear.

In any case, I think you have a bug in DirectCodeBuilder::buildContent.  
You either generate or reuse a stack map attribute, *and then you write 
all the attributes*. This looks like it may case a second SMA to be 
written in some cases.

The stack map attribute is similar to the bootstrap methods attribute; 
while it is an attribute, it is logically part of, and tightly coupled 
to, something else (for BMA, the constant pool.)  When we write the 
Class attributes, we explicitly filter the BMA; we should probably do so 
here -- instead of attributes.writeTo, we should iterate the attributes 
and write all but a SMA.


On 8/17/2022 8:54 AM, Adam Sotona wrote:
> ------------------------------------------------------------------------
>
>
>         You can view, comment on, or merge this pull request online at:
>
> https://github.com/openjdk/jdk-sandbox/pull/32
>
>
>         Commit Summary
>
>   * 84cfa01
>     <https://github.com/openjdk/jdk-sandbox/pull/32/commits/84cfa0103402266e8fbbccb8c0589c9180d8ea22>
>     removed FrameKind from StackMapTableAttribute and offsets replaced
>     with labels in StackMapFrames
>   * 456cc5e
>     <https://github.com/openjdk/jdk-sandbox/pull/32/commits/456cc5e9649a498f4c4d66465e458321cad72762>
>     simplification of StackMapFrame, removal of compressed sub-types
>   * c16382c
>     <https://github.com/openjdk/jdk-sandbox/pull/32/commits/c16382cc663528ecb08fa74c6220905d08cc1d31>
>     Implemented UnboundStackMapTableAttribute and added related
>     factory methods
>   * f507653
>     <https://github.com/openjdk/jdk-sandbox/pull/32/commits/f5076534bff609d130714571845ddea57f8a0352>
>     implemented StackMapTableAttribute compression and writing
>   * bfad2d1
>     <https://github.com/openjdk/jdk-sandbox/pull/32/commits/bfad2d11f5c4de4ac238e649cfe821cc083b415d>
>     StackMapTableAttribute extends CodeElement
>   * 5b4459d
>     <https://github.com/openjdk/jdk-sandbox/pull/32/commits/5b4459d891b5669f66fcb507b6cef7c36d6e9747>
>     added Classfile.Option.processStackMaps
>
>
>         File Changes
>
> (15 files <https://github.com/openjdk/jdk-sandbox/pull/32/files>)
>
>   * *M* src/java.base/share/classes/jdk/classfile/Attributes.java
>     <https://github.com/openjdk/jdk-sandbox/pull/32/files#diff-b8127cf9dcf2da7042ec3a390eac929ee3976c2c4bc31a1192ebd6e1c704f639>
>     (7)
>   * *M* src/java.base/share/classes/jdk/classfile/Classfile.java
>     <https://github.com/openjdk/jdk-sandbox/pull/32/files#diff-a06b37dbec06f035d014d970b71f079b7a68ab22efa44fbc2cfedf3f6cbb1f29>
>     (10)
>   * *M* src/java.base/share/classes/jdk/classfile/CodeElement.java
>     <https://github.com/openjdk/jdk-sandbox/pull/32/files#diff-e62b801db2681d114a264076d22b83afa008a8f85272c20b63e4b4f0ecb16dc5>
>     (6)
>   * *M* src/java.base/share/classes/jdk/classfile/Opcode.java
>     <https://github.com/openjdk/jdk-sandbox/pull/32/files#diff-025abd321469aaa4327bcfbb1af2cfbc68c36a89a66523e0be5e567c2acd9a5e>
>     (3)
>   * *M*
>     src/java.base/share/classes/jdk/classfile/attribute/StackMapTableAttribute.java
>     <https://github.com/openjdk/jdk-sandbox/pull/32/files#diff-593a350680c756f740b10b6f9542d56b827e5411fda6947eb4be7c35d9db2587>
>     (155)
>   * *M*
>     src/java.base/share/classes/jdk/classfile/impl/AttributeHolder.java
>     <https://github.com/openjdk/jdk-sandbox/pull/32/files#diff-7fb2303b71926c55ae4a6c82f95265d6abad48a7d4c9d862d3323b1d21d588a9>
>     (2)
>   * *M*
>     src/java.base/share/classes/jdk/classfile/impl/BoundAttribute.java
>     <https://github.com/openjdk/jdk-sandbox/pull/32/files#diff-d70dfa7e41a031dc96c57b40226b50102f3c42c33e6f98ca82794a28e146488c>
>     (32)
>   * *M*
>     src/java.base/share/classes/jdk/classfile/impl/ClassPrinterImpl.java
>     <https://github.com/openjdk/jdk-sandbox/pull/32/files#diff-90c3dcac096e2b20f5b3417844136b5e81ebeb19ae3e5cbb1e1dd05a76e2ae01>
>     (24)
>   * *M* src/java.base/share/classes/jdk/classfile/impl/CodeImpl.java
>     <https://github.com/openjdk/jdk-sandbox/pull/32/files#diff-396682fa6f06ca801296ca1c9823fa457d4633ceced358daf73722231c71e8d6>
>     (84)
>   * *M*
>     src/java.base/share/classes/jdk/classfile/impl/DirectCodeBuilder.java
>     <https://github.com/openjdk/jdk-sandbox/pull/32/files#diff-59eed26106daed5e712c101b4c643a8e1bf2e4c35a58195ace6d5187b2eee09f>
>     (6)
>   * *M* src/java.base/share/classes/jdk/classfile/impl/Options.java
>     <https://github.com/openjdk/jdk-sandbox/pull/32/files#diff-19c024f7ee4b14198f106204bc77ae2fd6d3779b31ebed3f05dce2293e1dc562>
>     (3)
>   * *M*
>     src/java.base/share/classes/jdk/classfile/impl/StackMapDecoder.java
>     <https://github.com/openjdk/jdk-sandbox/pull/32/files#diff-db6326200e0cf8d0ed5a77f3235bc9674f8bd05f1e5205a4a588977f4fa37584>
>     (299)
>   * *M*
>     src/java.base/share/classes/jdk/classfile/impl/UnboundAttribute.java
>     <https://github.com/openjdk/jdk-sandbox/pull/32/files#diff-234f798c2ece8ce169f5fcab0dd24953c37fdcdd4323c434f4a585c186c6517b>
>     (32)
>   * *M* test/jdk/jdk/classfile/helpers/RebuildingTransformation.java
>     <https://github.com/openjdk/jdk-sandbox/pull/32/files#diff-cfda863702717fe19b8ac97e8df5be7b7d54592df348b3ff3bfbc0edc3c235f0>
>     (15)
>   * *M* test/jdk/jdk/classfile/helpers/Transforms.java
>     <https://github.com/openjdk/jdk-sandbox/pull/32/files#diff-3a777cc5c56c6517fa74ff1df7f7c8f5b861dafcb0b5b1e7658f865cf94165c1>
>     (2)
>
>
>         Patch Links:
>
>   * https://github.com/openjdk/jdk-sandbox/pull/32.patch
>   * https://github.com/openjdk/jdk-sandbox/pull/32.diff
>
> —
> Reply to this email directly, view it on GitHub 
> <https://github.com/openjdk/jdk-sandbox/pull/32>, or unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/AABJ4RAHI6MNRZEHP4EKINLVZTOH3ANCNFSM56ZOCJ2Q>.
> You are receiving this because you are subscribed to this 
> thread.Message ID: ***@***.***>
>

asotona commented:
From: Brian Goetz ***@***.***>
Date: Tuesday, 11 October 2022 9:12
To: openjdk/jdk-sandbox ***@***.***>
Cc: Adam Sotona ***@***.***>, State change ***@***.***>
Subject: [External] : Re: [openjdk/jdk-sandbox] Classfile API stack map processing (PR #32)
In reviewing this PR, I noticed a vestige of an earlier abandoned
direction, which is CodeElement.Kind.  There are Kind values for
PARAMETER_ANNOTATION (basically unused), TYPE_ANNOTATION (misused), and
you added one for STACK_MAP.  I think all three of these should be
replaced with a single ATTRIBUTE kind, and the codeKind() of any
attribute that extends CodeElement should be ATTRIBUTE.  The END kind is
also effectively dead and should be removed (it is used only by a dead
opcode.)

OK, I’ll work on CodeElement.Kind removal.


This vestige is also present in your stack map attribute:

        @Override
        public Opcode opcode() {
            return Opcode.TYPE_ANNO;
        }

I cannot find this one.


which is clearly the wrong opcode to return for stack maps, but its also
the wrong thing to return for type annotation attributes as well.
Probably best to create a pseudo opcode for ATTRIBUTE to return in these
places too.

OK, will use common pseudo opcode ATTRIBUTE.

My reading of this PR is that:

 - You add a new option, PROCESS_STACK_MAP
 - The behavior of PSM is to deliver the StackMap attribute whole when
iterating the code elements (as we do with e.g. RVTA).
 - If SM generation has been disabled, and the user has send a stack
map attribute to the builder, we can write that (?)
Yes.


I like that you have to disengage the automatic stack map generation in
order to be able to generate your own, but this last bullet still makes
me a little uncomfortable.  I'll think about how to make this more clear.

It is an important corner-case (<5%), so we agreed it can be a bit uncomfortable.


In any case, I think you have a bug in DirectCodeBuilder::buildContent.
You either generate or reuse a stack map attribute, *and then you write
all the attributes*. This looks like it may case a second SMA to be
written in some cases.

If user provides custom SMA and does not disable generation/reuse – it is overridden.
AttributeHolder::withAttribute checks if AttributeMapper::allowMultiple and overrides the user provided SMA with the generated/reused one.


The stack map attribute is similar to the bootstrap methods attribute;
while it is an attribute, it is logically part of, and tightly coupled
to, something else (for BMA, the constant pool.)  When we write the
Class attributes, we explicitly filter the BMA; we should probably do so
here -- instead of attributes.writeTo, we should iterate the attributes
and write all but a SMA.

SMA has special treatment - it is not streamed. When SMA appears in attributes, it is always explicitly user-provided instance.
If user reads or transforms a class and needs to access original SMA, it is available in the model only (not in the stream).
User can manually pass original (or modified) SMA to the CodeBuilder and it will be applied only when SM generation is disabled.
No standard transformation will pass SMA through.
This was the latest output from our SMA-treatment discussions.

asotona commented:
I’ve prepared CodeElement.Kind cleanup pull request:
https://github.com/openjdk/jdk-sandbox/pull/41

  *   CodeElement.Kind moved to Opcode
  *   CodeElement::opcode and CodeElement::sizeInBytes moved down to Instruction
  *   removed all pseudo Opcodes and Kinds
Please review.

Thanks,
Adam

From: Adam Sotona ***@***.***>
Date: Tuesday, 11 October 2022 11:06
To: openjdk/jdk-sandbox ***@***.***>, openjdk/jdk-sandbox ***@***.***>
Cc: State change ***@***.***>
Subject: Re: [External] : Re: [openjdk/jdk-sandbox] Classfile API stack map processing (PR #32)


From: Brian Goetz ***@***.***>
Date: Tuesday, 11 October 2022 9:12
To: openjdk/jdk-sandbox ***@***.***>
Cc: Adam Sotona ***@***.***>, State change ***@***.***>
Subject: [External] : Re: [openjdk/jdk-sandbox] Classfile API stack map processing (PR #32)
In reviewing this PR, I noticed a vestige of an earlier abandoned
direction, which is CodeElement.Kind.  There are Kind values for
PARAMETER_ANNOTATION (basically unused), TYPE_ANNOTATION (misused), and
you added one for STACK_MAP.  I think all three of these should be
replaced with a single ATTRIBUTE kind, and the codeKind() of any
attribute that extends CodeElement should be ATTRIBUTE.  The END kind is
also effectively dead and should be removed (it is used only by a dead
opcode.)

OK, I’ll work on CodeElement.Kind removal.


This vestige is also present in your stack map attribute:

        @Override
        public Opcode opcode() {
            return Opcode.TYPE_ANNO;
        }

I cannot find this one.


which is clearly the wrong opcode to return for stack maps, but its also
the wrong thing to return for type annotation attributes as well.
Probably best to create a pseudo opcode for ATTRIBUTE to return in these
places too.


OK, will use common pseudo opcode ATTRIBUTE.

My reading of this PR is that:

 - You add a new option, PROCESS_STACK_MAP
 - The behavior of PSM is to deliver the StackMap attribute whole when
iterating the code elements (as we do with e.g. RVTA).
 - If SM generation has been disabled, and the user has send a stack
map attribute to the builder, we can write that (?)
Yes.


I like that you have to disengage the automatic stack map generation in
order to be able to generate your own, but this last bullet still makes
me a little uncomfortable.  I'll think about how to make this more clear.

It is an important corner-case (<5%), so we agreed it can be a bit uncomfortable.


In any case, I think you have a bug in DirectCodeBuilder::buildContent.
You either generate or reuse a stack map attribute, *and then you write
all the attributes*. This looks like it may case a second SMA to be
written in some cases.

If user provides custom SMA and does not disable generation/reuse – it is overridden.
AttributeHolder::withAttribute checks if AttributeMapper::allowMultiple and overrides the user provided SMA with the generated/reused one.


The stack map attribute is similar to the bootstrap methods attribute;
while it is an attribute, it is logically part of, and tightly coupled
to, something else (for BMA, the constant pool.)  When we write the
Class attributes, we explicitly filter the BMA; we should probably do so
here -- instead of attributes.writeTo, we should iterate the attributes
and write all but a SMA.

SMA has special treatment - it is not streamed. When SMA appears in attributes, it is always explicitly user-provided instance.
If user reads or transforms a class and needs to access original SMA, it is available in the model only (not in the stream).
User can manually pass original (or modified) SMA to the CodeBuilder and it will be applied only when SM generation is disabled.
No standard transformation will pass SMA through.
This was the latest output from our SMA-treatment discussions.
